add general container and entrypoint

itaysk opened the pull request:
close #463 
close #464
close #466

yanivagman commented on entrypoint.sh:
```diff
@@ -0,0 +1,32 @@
+	exit
+fi
+
+if [ "$1" = "--rules" ]; then
```
Why we use "--" here, unlike "trace"?
Also, shouldn't we accept any other flags than "--rules"?

itaysk replied:
> Why we use "--" here, unlike "trace"?

because trace is a subcommand, and this is a flag to the root command

> shouldn't we accept any other flags than "--rules"

yes eventually, this PR just satisfies the MVP scope so we can make a first appearance

yanivagman replied:
`because trace is a subcommand, and this is a flag to the root command`
Didn't we say that `rules` will be a subcommand just like `trace` is?
Also, if we choose to have a tracee-ebpf image like I suggested in #463 maybe we can just remove the `trace` subcommand.

itaysk replied:
> Didn't we say that rules will be a subcommand just like trace is?

I don't think so, maybe I missed something. but I don't see the point in adding a `rules` subcommand, and even if we had it, we would probably have made it the default

> Also, if we choose to have a tracee-ebpf

seems like we have swapped opinions :) initially I thought like you're suggesting (separate images), and then you suggested to use the same image with subcommand, and I think it a great idea, because we already have an image with all the tools inside, why not allow people to use those tools directly. 
We can still add another container dedicated for tracee-ebpf if you want to (I don't see why, it's just a few MB difference), but that doesn't have to mean we can't expose tracee-ebpf also from this one

yanivagman replied:
Yes, we can keep the `trace` subcommand.
I still think we need to have another image for tracee-ebpf as it is self-contained, and users who only need tracing (and forensics) capabilities, can just use it instead of the "tracee" image and have a simpler UX for their needs.
Moreover, the fact that today the difference is only just a few MB doesn't say that this will be the situation in the future - it will probably keep growing as we will add signatures and other functionality

itaysk replied:
ok.. I didn't say we can't have another image. this is a discussion on the entrypoint of `tracee`

itaysk commented on entrypoint.sh:
```diff
@@ -0,0 +1,32 @@
+	flag_webhook_url="--webhook=http://localhost:2801"
+fi
+
+$TRACEE_EBPF_EXE --output=format:gob --output=option:eot | $TRACEE_RULES_EXE --tracee-file=/dev/stdin $flag_webhook_url $flag_rules
```
did you read the PR message? I'm waiting for it to be merged then I'll make this ready.

Is there a problem with stdin? why would it not scale?

yanivagman replied:
`why would it not scale?`
As we already discussed, adding other sources will require the usage of pipes, and can't use stdin in parallel

`Is it something that would stop the MVP? Or can be addressed later?`
Doesn't have to stop MVP, but it feels like a regression to me.
We had file input (not stdin), which could have supported streaming input source before the removal of eot.
Now we took step back - and only stdin supports streaming

itaysk replied:
> adding other sources will require the usage of pipes, and can't use stdin in parallel

oh ok.. but we have just one input source now, and file is not a viable option (see below)

> We had file input (not stdin), which could have supported streaming input source before the removal of eot.
Now we took step back - and only stdin supports streaming

I don't think it's a step back because the file based solution was practically unusable (file would indefinitely grow with no way to trim or partition it).

Completely agree that the best way forward is something like named pipes, but we want to release soon and for the sake of the MVP, I think this can be good enough.

yanivagman replied:
Agree. For MVP that's enough.

> I don't think it's a step back because the file based solution was practically unusable (file would indefinitely grow with no way to trim or partition it).

File doesn't have to grow indefinetly. it's maximum size can be specified and drop events when full (like a circular buffer). But I don't think that going forward this is the correct solution anyways, I just wanted to make the point that using stdin would not scale

grantseltzer commented on entrypoint.sh:
```diff
@@ -0,0 +1,32 @@
+#!/bin/sh
```
We should set pipefail in our bash scripts:`set -euf -o pipefail`

([ref](https://sipb.mit.edu/doc/safe-shell/))

itaysk replied:
as a general rule of thimb I guess this is true, but in this case:
- `pipefail` is not portable
- `-u` will break checks for optional arguments (which may be legitimately unset)
- `-e` will break checking for non-zero exit codes (like grep didn't find or test eval to false)

maybe there's a way to rewrite the entire script to work in strict mode but I'm not sure it's worth the effort

simar7 commented:
Just poking stuff but how do I run this? And how do I exit? A simple SIGINT doesn't seem to have an effect.

```
vagrant@ubuntu-groovy:/usr/vagrant/repos/tracee$ docker run itaysk:entry
mount: permission denied (are you root?)
2021/02/08 19:26:14 error opening plugin builtin.so: plugin.Open("/tracee/rules/builtin.so"): Error loading shared library /tracee/rules/builtin.so: Exec format error
2021/02/08 19:26:14 [INFO]  : Enabled Outputs : 
2021/02/08 19:26:14 [INFO]  : Falco Sidekick is up and listening on port 2801
2021/02/08 19:26:14 missing kernel source code compilation dependency
^C^C^C
```

itaysk commented:
@simar7 check out the tracee-ebpf readme, try something like:
```
docker run  --rm --privileged --pid=host -v /lib/modules/:/lib/modules/:ro -v /usr/src:/usr/src:ro itaysk:entry 
```

I'm looking into the SIGINT issue

itaysk commented:
@simar7 I need more info about the SIGINT issue you're having not sure I can repro

simar7 commented on Dockerfile:
```diff
@@ -0,0 +1,28 @@
+WORKDIR /tracee
+
+COPY --from=falcosecurity/falcosidekick@sha256:a32d8850d51e9b096a09f4ae73ba6cde038c3fe1fd9c58baf76333dfda7e7bbd /app/falcosidekick ./
+COPY --from=build /tracee/tracee-ebpf/dist/tracee-ebpf /tracee/tracee-rules/dist/tracee-rules /tracee/entrypoint.sh ./
+COPY --from=build /tracee/tracee-rules/dist/rules ./rules
```
Should we rebuild these artifacts before we copy?

itaysk replied:
good catch. I'll fix it

itaysk commented:
> @simar7 I need more info about the SIGINT issue you're having not sure I can repro

@simar7 I think what's happening is that you didn't attach TTY, you are just looking at it's stdout. try adding `-it` to docker run

simar7 commented:
> > @simar7 I need more info about the SIGINT issue you're having not sure I can repro
> 
> @simar7 I think what's happening is that you are running in foreground but didn't attach TTY, you are just looking at it's stdout. try adding `-it` to docker run

Ah that would explain. But I'm running into another issue:

```
vagrant@ubuntu-groovy:/usr/vagrant/repos/tracee$ docker run -it  --rm --privileged --pid=host -v /lib/modules/:/lib/modules/:ro -v /usr/src:/usr/src:ro itaysk:entry
Error relocating /tracee/tracee-rules: __vfprintf_chk: symbol not found
Error relocating /tracee/tracee-rules: __fprintf_chk: symbol not found
2021/02/08 19:43:40 [INFO]  : Enabled Outputs : 
2021/02/08 19:43:40 [INFO]  : Falco Sidekick is up and listening on port 2801
failed to add kprobe 'p:kprobes/psecurity_inode_unlink security_inode_unlink': -17
failed to create kprobe event: -17
```

itaysk commented:
looks like your container is packing execs that were build on your host using glibc. how did you build the image?

itaysk commented:
just to be sure, delete all `dist` directories from the root and from `tracee-*` and `docker build -t something .` in the root

simar7 commented:
> looks like your container is packing execs that were build on your host using glibc. how did you build the image?

To build this image:
```
docker build . -t itaysk:entry 
```



> looks like your container is packing execs that were build on your host using glibc. how did you build the image?


I removed all `dist` dirs and rebuilt both tracee-ebpf and tracee-rules prior to building this docker image. For building the first two I just ran `make all` in their respective directories and to build the image I ran `docker build --no-cache -t itaysk:entry .`

Here's the most recent output:
```
docker run -it  --rm --privileged --pid=host -v /lib/modules/:/lib/modules/:ro -v /usr/src:/usr/src:ro itaysk:entry
Error relocating /tracee/tracee-rules: __vfprintf_chk: symbol not found
Error relocating /tracee/tracee-rules: __fprintf_chk: symbol not found
Error relocating /tracee/tracee-ebpf: __snprintf_chk: symbol not found
Error relocating /tracee/tracee-ebpf: __vfprintf_chk: symbol not found
Error relocating /tracee/tracee-ebpf: __memcpy_chk: symbol not found
Error relocating /tracee/tracee-ebpf: __strcpy_chk: symbol not found
Error relocating /tracee/tracee-ebpf: __sprintf_chk: symbol not found
```

I'm building on a Ubuntu 20.10 FWIW, I never ran into issues running tracee-ebpf or tracee-rules on their own.

yanivagman commented:
@simar7 note that we have two dockerfiles - Dockerfile and Dockerfile,builder.
The first Dockerfile is using alpine (musl libc), and is what we use while doing a release.
The second Dockerfile.builder is using debian base image (glibc) and is what used for local builds.
It may be that you mixed the two in some way - built locally the executables (using glibc), then build the images using the first Dockerfile

itaysk commented:
> rebuilt both tracee-ebpf and tracee-rules prior to building this docker image

don't do this. just clean all dists and run `docker build -t .` that's it

simar7 commented:
> > rebuilt both tracee-ebpf and tracee-rules prior to building this docker image
> 
> don't do this. just clean all dists and run `docker build -t .` that's it

Thanks that did the trick. I think we should document how to use the entrypoint/image with this PR. Thoughts?

What if we clean the respective `dist/` directories when we run this entry point? Or is there any other way to ignore the current artifacts present within them to build the entry point image?

itaysk commented:
> I think we should document how to use the entrypoint/image

yes absolutely

> What if we clean the respective dist/ directories when we run this entry point? Or is there any other way to ignore the current artifacts present within them to build the entry point image?

I'll try to think of something

simar7 commented:
> > I think we should document how to use the entrypoint/image
> 
> yes absolutely

It could be just as simple as the following (I didn't know where this should live to I'm writing it down here):

### Tracee Runner Image
In order to facilitate ease of use, we've created a general Tracee container that contains ready to use release binaries for tracee-ebpf and tracee-rules. It also contains an instance of Falco sidekick to facilitate webhook events.

To run this container, simply: 
`make build-tracee-runner` which runs the following docker command
`docker build -t aquasec/tracee-runner:latest` to build a docker image.

This will do the following:
1. Build tracee-ebpf and tracee-rules binaries.
2. Copy the above built as artifacts into the new runner container image.
3. Copy Falco sidekick (https://github.com/falcosecurity/falcosidekick) for Webhook delivery.

The runner image needs to run in docker privileged mode. A simple invocation is like the following:
```docker run -it  --rm --privileged --pid=host -v /lib/modules/:/lib/modules/:ro -v /usr/src:/usr/src:ro aquasec/tracee-runner```

Or by invoking the Makefile target: `make run-tracee-runner`

Note here that I made the following assumptions:
1. We call this orchestration/demo container as "tracee-runner". I'm happy with another name. Also feel free to change the make target names.
2. Added a bunch of Makefile targets to run the docker commands we need to build.

simar7 commented on entrypoint.sh:
```diff
@@ -0,0 +1,32 @@
+
+TRACEE_EBPF_EXE=${TRACEE_EBPF_EXE:="/tracee/tracee-ebpf"}
+TRACEE_MOUNT_EXE=${TRACEE_MOUNT_EXE:="mount"}
+TRACEE_WEBHOOK_CONFIG=${TRACEE_WEBHOOK_CONFIG:="/tmp/tracee/integrations-config.yaml"}
```
For my understanding: How is this used? I see Falco sidekick has a WebUI that I was expecting to see if I didn't pass in any arguments https://github.com/falcosecurity/falcosidekick-ui and https://github.com/falcosecurity/falcosidekick/blob/master/config_example.yaml#L164-L165

I'm just trying to configure Falco sidekick to be able to look at the events that Tracee will generate.

itaysk replied:
IDK what's webUI, seems like a new feature added to master 2 days ago and yet unreleased

itaysk replied:
I just configured it with MS Teams output to see that it works

itaysk commented on Readme.md:
```diff
@@ -9,9 +9,95 @@
+Alternatively, you can pre-compile the eBPF program, and provide it to Tracee. There are some benefits to this approach as you will not need clang and kernel headers at runtime anymore, as well as reduced risk of invoking an external program at runtime.
+
+You can build the eBPF program in the following ways:
+1. Clone the repo and `make bpf`.
+2. `make bpf DOCKER=1` to build in a Docker container which includes all development tooling.
```
I'll add this in a future PR. don't want to make this PR larger then it already is

itaysk replied:
#519

itaysk commented:
added readme to the root level. it should include all of the operational instructions that applies to tracee-ebpf so I copied most of it in. it's Once we complete #435 (post-MVP) we will refactor the docs anyway.
also added `--list` flag to the container as a way for users to see what rules are provided, until we have AVD integration

yanivagman commented on Makefile:
```diff
@@ -1,14 +1,12 @@
 OUT_CHECKSUMS := $(OUT_DIR)/checksums.txt
 PUSH_DOCKER_REPO ?= aquasec/tracee
 PUSH_DOCKER_TAG ?= $(RELEASE_TAG:v%=%)
+RELEASE_FILES := LICENSE $(OUT_DIR)/tracee-ebpf $(OUT_DIR)/tracee-rules $(OUT_DIR)/rules
+# RELEASE_TAG must be set for the release target
+VERSION ?= $(if $(RELEASE_TAG),$(RELEASE_TAG),$(shell $(CMD_GIT) describe --tags 2>/dev/null || echo '0'))
```
Can we now remove these same lines from tracee-ebpf/Makefile and other similar stuff as well?

itaysk replied:
no because it's used to determine the name of the bpf obj (tracee.bpf.linuxversion.traceeversion.o)

itaysk replied:
but we could clean up the tracee-ebpf makefile depending on what we decide to do with #255 (I'm leaving it to that PR)
